File: scripts/example3.m
% script: example3.m
% This script demonstrates how pre-polarization switch-off ramps are simulated
% with BLOCHUS.
% Both examples were also used as benchmark cases in the paper:
%	Hiller, T., Dlugosch, R. and Müller-Petke, M., "Utilizing pre-
%	polarization to enhance SNMR signals - effect of imperfect
%	switch-off", Geophysical Journal International Vol. 222(2), p.815-826, 2020
%
% Two pre-polarization switch-off benchmarks are shown:
%
% 1.) from Melton et al., 1995, J Mag Res A, Vol. 117, p.164-170
%
%     general settings:
%     - magnetic field B0 is set to 50 µT
%     - pre-polarization field is a factor 100 larger
%     - uses a linear switch-off ramp with different switch-off rate (slopes)
%
% 2.) Conradi et al., 2017, J Mag Res, Vol. 281, p.241-245
%
%     general settings:
%     - magnetic field B0 is set to 50 µT
%     - pre-polarization field is a factor 50 larger
%     - uses a linexp switch-off ramp with different initial angles theta
%     (angle between B0 and Bp) and different switch-over fields Bstar
%
% See also: BLOCHUS
% Author: Thomas Hiller
% email: thomas.hiller[at]leibniz-liag.de
% License: GNU GPLv3 (at end)

%------------- BEGIN CODE --------------

% clear data
clear variables; clc;

% activate different tests
% 1 - Melton et al. 1995 benchmark
% 2 - Conradi et al. 2017 benchmark

% because this script takes a few minutes ask the user before continuing
answer = questdlg({'This script needs about 5 min. to run.',...
    'Do you want to continue?'},'Dessert Menu','Yes','No','Yes');
% handle response
switch answer
    case 'Yes'
        usetest = [1 1];
    case {'No',''}
        usetest = [0 0];
end

if sum(usetest) > 0
    % basic settings
    zunit = [0 0 1]'; % z unit vector
    % ODE solver error tolerance
    tol = 1e-9;
    % ODE solver options
    options = odeset('RelTol',tol,'AbsTol',[tol tol tol]);
end
%% 1.) Melton et al.(1995)
if usetest(1)
    % hydrogen proton
    nucleus = '1H';
    % pre-polarization factor [B0]
    PrePolFactor = 100;
    
    % parameter needed for the ODE solver
    % simulation type [string]
    odeparam.type = 'prepol';
    % equilibrium magnetization [A/m]
    odeparam.M0 = zunit;
    % primary (Earth's) magnetic field B0 [T]
    odeparam.B0 = 5e-5;
    % relaxation time T1 [s]
    odeparam.T1 = 2;
    % relaxation time T2 [s]
    odeparam.T2 = 1;
    % gyromagnetic ratio [rad/s/T]
    odeparam.gamma = getGyroRatio(nucleus);
    
    % pre-polarization switch-off ramp parameter
    % relaxation during switch-off [0/1]
    odeparam.RDS = 0;
    % switch-off ramp type [string]
    rampparam.ramp = 'lin';
    % amplitude of pre-polarization field
    rampparam.Bmax = odeparam.B0*PrePolFactor;
    % switch over magnetization (for linexp case)
    rampparam.Bstar = odeparam.B0;
    
    % initial orientation is toards y-axis
    orient = getRotationMatrixFromAngleandAxis(pi/2,[0 1 0])*zunit;
    orient = getRotationMatrixFromAngleandAxis(pi/2,[0 0 1])*orient;
    % initial magnetization in the direction of B0+Bp
    Morient = orient*rampparam.Bmax + odeparam.B0*zunit;
    % Minit does not get normalized in this case
    Minit = Morient;
    % because Minit is not normalized to 1, M0 needs to be adjusted
    odeparam.M0 = odeparam.B0*zunit;
    
    % orientation of the pre-polarization pulse axis (y-axis)
    odeparam.orient = orient;
    
    % predefined relaxation rates (k/GAMMA)
    swrates = [1/16 1/8 1/4 1/2 1 2 4 8 16];
    % initialize output variables
    M_final = zeros(numel(swrates),3) ;
    Phi_final = zeros(numel(swrates),1) ;
    Theta_final = zeros(numel(swrates),1) ;
    for r = 1:numel(swrates)
        
        % current relaxation rate
        rate = swrates(r);
        % calculate switch-off ramp time
        GAMMA = PrePolFactor/rate;
        Tramp = GAMMA/(odeparam.B0*odeparam.gamma);
        
        % switch-off ramp time [s]
        rampparam.Tramp = Tramp;
        % switch over ramp time [s] (for linexp case)
        rampparam.Tslope = Tramp;
        
        % add the ramp parameter to the ode parameter struct
        odeparam.rampparam = rampparam;
        
        % ODE solver call
        % OUTPUT: time T and magnetization M
        [T,M] = ode45(@(t,m) fcn_BLOCHUS_ode(t,m,odeparam),[0 Tramp],Minit,options);
        
        % calculate final angles describing the orientation of M
        % see p. 167 of the paper
        theta = 90 - acosd(M(end,3)/norm(M(end,:)));
        phi = 90 - atan2d(M(end,2),M(end,1));
        
        % output data
        M_final(r,:) = M(end,:)./odeparam.B0;
        Phi_final(r) = phi;
        Theta_final(r) = theta;
    end
    
    f1 = figure; set(f1,'Name','example3a_ref');
    % color the different switch-off rates
    col = jet(numel(swrates));
    ax1 = subplot(1,2,1);
    hold(ax1,'on');
    % plot a colored arrow showing the endpoint of the magnetization
    % this corresponds to Fig. 3 in Melton et al.(1995)
    for r = 1:numel(swrates)
        % normalize the arrow to 1
        M_final(r,:) = M_final(r,:)./norm(M_final(r,:));
        quiver3(0,0,0,M_final(r,1),M_final(r,2),M_final(r,3),0,...
            'Color',col(r,:),'LineWidth',2);
    end
    % plot the trace on the sphere surface
    plot3([0; M_final(:,1); 0],[0; M_final(:,2); 1],[1; M_final(:,3); 0],...
        'k--','LineWidth',2);
    % plot the Bloch sphere
    bsh = plotBSphere(30,30,ax1);
    text(-0.1,0.45,0.8,'30°','Parent',ax1);
    text(-0.1,0.8,0.4,'60°','Parent',ax1);
    set(ax1,'Color','w','XColor','none','YColor','none','ZColor','none');
    axis equal tight
    view([135 45]);
    set(ax1,'XLim',[0 1.2],'YLim',[0 1.2],'ZLim',[0 1.2]);
    
    % final angles for all switch-off rates
    % this corresponds to Fig. 4 in Melton et al.(1995)
    ax2 = subplot(1,2,2);
    hold(ax2,'on');
    for r = 1:numel(swrates)
        plot(swrates(r),Phi_final(r),'s','Color',col(r,:),'MarkerSize',8);
        plot(swrates(r),Theta_final(r),'o','Color',col(r,:),'MarkerSize',8);
    end
    % the fit from the paper eq.7
    xI = 0.5:16; yI = 50.8./sqrt(xI);
    plot(xI,yI,'k--');
    x = [0.675 0.775];
    y = [0.4 0.5];
    ah = annotation('textarrow',x,y,'String','$$50.8^\circ \sqrt(k/\Gamma)$$');
    set(ah,'Interpreter','latex','TextBackgroundColor','w','TextLineWidth',1);
    set(ax2,'XScale','log','YScale','log');
    set(ax2,'XLim',[0.04 30],'XTick',[0.1 1 10],'XTickLabel',{'0.1','1','10'})
    set(ax2,'YLim',[10 100])
    grid on; box on; axis square;
    legend('\phi_f','\theta_f');
    xlabel('relaxation rate k/\Gamma');ylabel('final angles [°]');
    
end

%% 2.) Conradi et al.(2017)
if usetest(2)
    
    % Fig. 4a of Conradi et al.(2017)
    %     conradi_Bmax = 16/1e3; % [T]
    %     conradi_Bstar = [2000 250 120]./1e6; % [T]
    %     conradi_Tslope = 6/1e3; % [s]
    %     conradi_lgdstr = {'2000','250','120'};
    
    % Fig. 4b of Conradi et al.(2017)
    conradi_Bmax = 2.5/1e3; % [T]
    conradi_Bstar = [250 60 1]./1e6; % [T]
    conradi_Tslope = 10/1e3; % [s]
    conradi_lgdstr = {'250','6','1'};
    
    % hydrogen proton
    nucleus = '1H';
    % pre-polarization factor [B0]
    PrePolFactor = 50;
    
    % parameter needed for the ODE solver
    % simulation type [string]
    odeparam.type = 'prepol';
    % equilibrium magnetization [A/m]
    odeparam.M0 = zunit;
    % primary (Earth's) magnetic field B0 [T]
    odeparam.B0 = 5e-5;
    % relaxation time T1 [s]
    odeparam.T1 = 2;
    % relaxation time T2 [s]
    odeparam.T2 = 1;
    % gyromagnetic ratio [rad/s/T]
    odeparam.gamma = getGyroRatio(nucleus);
    
    % pre-polarization switch-off ramp parameter
    % relaxation during switch-off [0/1]
    odeparam.RDS = 0;
    % switch-off ramp type [string]
    rampparam.ramp = 'linexp';
    % amplitude of pre-polarization field [T]
    rampparam.Bmax = conradi_Bmax;
    
    % switch-off ramp time [s]
    rampparam.Tramp = conradi_Tslope*2;
    % switch over ramp time [s] (for linexp case)
    rampparam.Tslope = conradi_Tslope;
    
    % because Minit is not normalized to 1, M0 needs to be adjusted
    odeparam.M0 = odeparam.B0*zunit;
    
    % predefined angles theta [deg]
    theta = linspace(0,200,101);
    % predefined switch over magnetization [T]
    Bstar = conradi_Bstar;
    % initialize output variable
    p = zeros(numel(theta),numel(Bstar));
    for b = 1:numel(Bstar)
        for r = 1:numel(theta)
            
            % initial orientation is towards y-axis
            orient = getRotationMatrixFromAngleandAxis(deg2rad(theta(r)),[0 1 0])*zunit;
            % initial magnetization in the direction of B0+Bp
            Morient = orient*rampparam.Bmax + odeparam.B0*zunit;
            % Minit does not get normalized in this case
            Minit = Morient;
            % orientation of the pre-polarization pulse axis (y-axis)
            odeparam.orient = orient;
            
            % switch over magnetization (for linexp case) [T]
            rampparam.Bstar = Bstar(b);
            
            % add the ramp parameter to the ode parameter struct
            odeparam.rampparam = rampparam;
            
            % ODE solver call
            % OUTPUT: time T and magnetization M
            [T,M] = ode45(@(t,m) fcn_BLOCHUS_ode(t,m,odeparam),[0 rampparam.Tramp],Minit,options);
            
            % normalized magnetization vector at end of switch-off ramp
            MMn = M(end,:)./norm(M(end,:));
            % "adiabatic quality" p of the switch-off ramp
            % -> orientation of M with respect to z_unit
            p(r,b) = dot(MMn(end,:),zunit)./norm(zunit);
            
            clc;
            disp(['Bstar: ',sprintf('%d',Bstar(b)*1e6),'µT',...
                ' Theta: ',sprintf('%d',theta(r)),'°']);
        end
    end
    
    %% this corresponds to Fig. 4b in Conradi et al.(2017)
    f2 = figure; set(f2,'Name','example3b_ref');
    col = [1 0 0;0 1 0;0 0 1];
    ax1 = axes;
    hold(ax1,'on');
    for b = 1:numel(Bstar)
        plot(theta,p(:,b),'Color',col(b,:));
    end
    set(ax1,'XLim',[-20 210],'XTick',0:45:180);
    set(ax1,'YLim',[-1.05 1.05]);
    grid on; box on;
    lh = legend(conradi_lgdstr,'Location','SouthWest');
    set(get(lh,'Title'),'String','B^\ast [µT]')
    xlabel('angle \theta (\angle B_0 B_p) [°]');
    ylabel('adiabatic quality p');
    
end

%------------- END OF CODE --------------

%% License:
% GNU GPLv3
%
% BLOCHUS
% Copyright (C) 2019 Thomas Hiller
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <https://www.gnu.org/licenses/>.

File: pyBLOCHUS/examples/example3.py
"""Create example3 plots.

This script demonstrates how p-polarization switch-off ramps are simulated
with BLOCHUS.
Both examples were also used as benchmark cases in the paper:
Hiller, T., Dlugosch, R. and Müller-Petke, M., "Utilizing pre-
polarization to enhance SNMR signals - effect of imperfect
switch-off", Geophysical Journal International Vol. 222(2), p.815-826, 2020

Two p-polarization switch-off benchmarks are shown:

1.) from Melton et al., 1995, J Mag Res A, Vol. 117, p.164-170
    general settings:
     - magnetic field B0 is set to 50 µT
     - pre-polarization field is a factor 100 larger
     - uses a linear switch-off ramp with different switch-off rate (slopes)

2.) Conradi et al., 2017, J Mag Res, Vol. 281, p.241-245
    general settings:
        - magnetic field B0 is set to 50 µT
        - pre-polarization field is a factor 50 larger
        - uses a linexp switch-off ramp with different initial angles theta
          (angle between B0 and Bp) and different switch-over fields Bstar

See also: BLOCHUS
Author: Thomas Hiller
email: thomas.hiller[at]leibniz-liag.de
License: GNU GPLv3 (at end)
"""
import numpy as np
# we need several matplotlib libs for plotting
import matplotlib as mp
import matplotlib.pyplot as plt
import matplotlib.colors as colors
import matplotlib.cm as cmx
# although pylint is complaining, that these classes are not used (which
# technically is true), they have to be imported to make the 3d plots
from mpl_toolkits.mplot3d import axes3d, Axes3D  # pylint: disable=W0611

from pyBLOCHUS import BlochusPrepol as prepol

# activate different tests
# 1 - Melton et al. 1995 benchmark - NOTE: takes ~1min
# 2 - Conradi et al. 2017 benchmark - NOTE: takes ~7min
usetest = [True, True]
SAVEPNG = False


def test_one(save_png=False):
    """Perform the first test: 'Melton et al. 1995' benchmark."""
    # initialize prepol Bloch class
    bloch = prepol()
    # the ramp shape is linear -> 'lin'
    bloch.Ramp.shape = 'lin'
    # initial orientation is towards y-axis (phi=90
    bloch.Ramp.phi = 90
    # and in the x-y-plane (theta=90)
    bloch.Ramp.theta = 90

    # 'Ramp.orient' gets updated automatically
    # pre-polarization field into the direction of 'Ramp.orient'
    # and Ramp.factor x B0 stronger than B0
    b_p = bloch.Ramp.orient*bloch.Ramp.factor*bloch.B0
    # Earth's magnetic field (as vector)
    b_e = bloch.B0*bloch.zunit
    # initial magnetization points into the sum of both
    m_init = b_p + b_e
    # for simplicity equilibrium magnetization is B0 (obtaining the real
    # magnetization values is simply a multiplication of the magnetic fields
    # with the "Curie" factor)
    bloch.M0 = bloch.B0*bloch.zunit

    # predefined relaxation rates (k/gamma) from Melton et al. 1995
    relax_rates = np.array([1/16, 1/8, 1/4, 1/2, 1.0, 2.0, 4.0, 8.0, 16.0])
    # initialize output variables
    m_final = np.zeros((3, len(relax_rates)), dtype=float)
    phi_final = np.zeros((len(relax_rates),), dtype=float)
    theta_final = np.zeros((len(relax_rates),), dtype=float)
    # loop over relaxation rates
    print('start calculation of relaxation rates ...')
    for i in np.arange(0, len(relax_rates)):
        # current relaxation rate
        rate = relax_rates[i]
        # calculate switch-off ramp time
        gamma = bloch.Ramp.factor/rate
        bloch.Ramp.t_ramp = gamma/(bloch.B0*bloch.gamma)
        # update t_slope
        bloch.Ramp.t_slope = bloch.Ramp.t_ramp
        # update simulation time t_sim
        bloch.t_sim = bloch.Ramp.t_ramp
        # solve
        bloch.solve(m_init, use_numba=True)

        # calculate final angles describing the orientation of M
        # see p. 167 of the paper
        m_end = bloch.m[:, -1]
        m_end_n = np.linalg.norm(m_end)
        theta = 90 - np.rad2deg(np.arccos(m_end[2]/m_end_n))
        phi = 90 - np.rad2deg(np.arctan2(m_end[1], m_end[0]))

        # gather output data
        m_final[:, i] = m_end/m_end_n
        phi_final[i] = phi
        theta_final[i] = theta
        # inform the user about the progress
        print('{}/{} done'.format(i+1, len(relax_rates)))

    # create output figure
    fig = plt.figure(figsize=(9, 5))
    ax1 = fig.add_subplot(121, projection='3d')
    ax2 = fig.add_subplot(122)

    # draw only a part of the Bloch sphere
    bloch.plot_bloch_sphere(ax1, (30, 30), 1, (0, 90), (0, 90))

    # create an index vector for discrete colors out of the jet colormap
    index = range(len(relax_rates))
    jet = plt.get_cmap('jet')
    c_norm = colors.Normalize(vmin=0, vmax=index[-1])
    scalar_map = cmx.ScalarMappable(norm=c_norm, cmap=jet)

    # loop over all relaxation rates and plot the result with the
    # corresponding color
    for i in np.arange(0, len(relax_rates)):
        color_val = scalar_map.to_rgba(index[i])
        ax1.quiver(0, 0, 0, m_final[0, i], m_final[1, i], m_final[2, i],
                   color=color_val)
        ax2.plot(relax_rates[i], theta_final[i], marker='o', fillstyle='none',
                 color=color_val)
        ax2.plot(relax_rates[i], phi_final[i], marker='s', fillstyle='none',
                 color=color_val)

    # plot the trace of all final positions on the Bloch sphere surface
    m_x = np.append(np.append(np.array([0.0]), m_final[0, :]), np.array([0.0]))
    m_y = np.append(np.append(np.array([0.0]), m_final[1, :]), np.array([1.0]))
    m_z = np.append(np.append(np.array([1.0]), m_final[2, :]), np.array([0.0]))
    ax1.plot3D(m_x, m_y, m_z, 'k--')
    # adjust axis position
    ax1.set_position([0.0, -0.11, 0.35*1.5, 0.77*1.4])

    # plot the fit from the Melton et al. paper (eq. 7)
    x_1 = relax_rates[3:len(relax_rates)]
    y_1 = 50.8/np.sqrt(x_1)
    ax2.plot(x_1, y_1, 'k--')
    ax2.annotate(r'$50.8^{\circ}\sqrt{k/\Gamma}$', xy=(2.8, 30),
                 xycoords='data', xytext=(0.3, 20), textcoords='data',
                 arrowprops=dict(arrowstyle="->", connectionstyle="arc3"))

    # add a legend
    lgd_labels = [r'$\theta_f$', r'$\phi_f$']
    ax2.legend(lgd_labels, loc='upper right')

    # set axis properties
    ax2.set_xscale('log')
    ax2.set_yscale('log')
    ax2.set_yticks([10, 100])
    ax2.get_yaxis().set_minor_formatter(mp.ticker.NullFormatter())
    ax2.set_xlim(0.04, 30)
    ax2.set_ylim(10, 100)
    ax2.grid(color='lightgray', which='both', axis='both')
    ax2.set_xlabel(r'relaxation rate k/$\Gamma$')
    ax2.set_ylabel('final angles [deg]')

    # save figure
    if save_png:
        fig.savefig('example3a.png', dpi=300)


def test_two(save_png=False):
    """Perform the second test: 'Conradi et al. 2017' benchmark."""
    # choose which parameters should be used
    conradi_fig = '4b'
    if conradi_fig == '4a':
        conradi_factor = 320
        conradi_switch = np.array([40.0, 5.0, 2.4])
        conradi_t_slope = 0.006
        lgd_labels = ['2000', '250', '120']
    else:
        conradi_factor = 50
        conradi_switch = np.array([5.0, 1.2, 0.02])
        conradi_t_slope = 0.01
        lgd_labels = ['250', '6', '1']

    # initialize prepol Bloch class
    bloch = prepol()
    # the ramp shape is linear & exponential -> 'linexp'
    bloch.Ramp.shape = 'linexp'
    # the factor is 50
    bloch.Ramp.factor = conradi_factor
    # initial orientation is towards y-axis (phi=90)
    bloch.Ramp.phi = 90

    bloch.Ramp.t_slope = conradi_t_slope
    bloch.Ramp.t_ramp = bloch.Ramp.t_slope * 2
    bloch.t_sim = bloch.Ramp.t_ramp

    bloch.M0 = bloch.B0*bloch.zunit

    theta = np.linspace(0, 200, 101, endpoint=True)
    adiab_quality = np.zeros((len(theta), 3))

    for i in np.arange(0, len(conradi_switch)):
        bloch.Ramp.switch_factor = conradi_switch[i]
        for j in np.arange(0, len(theta)):
            # running theta value
            bloch.Ramp.theta = theta[j]

            # 'Ramp.orient' gets updated automatically
            # pre-polarization field into the direction of 'Ramp.orient'
            # and Ramp.factor x B0 stronger than B0
            b_p = bloch.Ramp.orient*bloch.Ramp.factor*bloch.B0
            # Earth's magnetic field (as vector)
            b_e = bloch.B0*bloch.zunit
            # initial magnetization points into the sum of both
            m_init = b_p + b_e
            # solve
            bloch.solve(m_init, use_numba=True)

            adiab_quality[j, i] = bloch.Ramp.adiab_qual
            # inform the user about the progress
            print('B* = {} [µT] | Theta = {} [deg]'.format(lgd_labels[i],
                                                           theta[j]))

    # create output figure
    fig = plt.figure(figsize=(7, 5))
    ax1 = fig.add_subplot(111)
    col = ['r', 'g', 'b']
    for i in np.arange(0, len(conradi_switch)):
        ax1.plot(theta, adiab_quality[:, i], color=col[i])

    ax1.set_xlim(-20, 210)
    ax1.set_xticks([0, 45, 90, 135, 180])
    ax1.set_ylim(-1.05, 1.05)
    ax1.set_yticks([-1, -0.5, 0, 0.5, 1])
    ax1.grid(color='lightgray', axis='both')
    ax1.set_xlabel(r'angle $\theta$ ($\angle B_0 B_p$) [deg]')
    ax1.set_ylabel(r'adiabatic quality $p$')
    ax1.legend(lgd_labels, title=r'$B^\ast$ [µT]', loc='lower left')

    # save figure
    if save_png:
        fig.savefig('example3b.png', dpi=300)


# actually call the main function
# ask the user to proceed
QSTR = 'This script runs approx. 8min on a (Core i5-6200) Laptop. '
QSTR += 'Continue? [y/n]: '
answer = input(QSTR)
if answer in {"y", "ye", "yes"}:
    if usetest[0]:
        test_one(SAVEPNG)
    if usetest[1]:
        test_two(SAVEPNG)
else:
    print("Canceled. Bye.")


# License:
# GNU GPLv3
#
# blochUS
# Copyright (C) 2019 Thomas Hiller
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
